### ISSUE TITLE
Separate read and write instances using the same storage backend

### ISSUE THREAD

**Issue by sagikazarmark** [distribution/distribution#3573]
I'd like to set up a registry that is read-only for the outside world and writable from a certain set of places.

My plan is the following:
- Run a set of read-only instance accessible from the internet
- Run a writeable instance accessible only from private network

The different instances would use the same storage backend, but could use separate auth backends. DNS is not that important, but let's say they run on different hosts for the sake of simplicity.

Is this kind of setup feasible with distribution? I'm certainly going to try it myself, so I'm mostly interested in the potential "hidden" issues (like garbage collection running in different instances causing trouble, different host being an issue or anything else that wouldn't occur immediately).

This would be an additional layer of security (in addition to the custom auth backend that would only issue pull credentials, but would also save the hassle of being able to issue both types of credentials), but I'm not sure it's worth it. Anyone running registries in this setup?

Thanks in advance!

(Sorry if this topic was already discussed, a 15 minutes search didn't yield any results)

**Comment by milosgajdos** [comment]
`distribution` provides a `read-only` mode configuration option which is rather confusingly "hidden" under `maintenance" https://docs.docker.com/registry/configuration/#readonly.

Alternatively, you could also run the edge `distribution` instances in `pull-through` cache mode, though mind you, there are some outstanding issues that should be addressed in the proxy code (please browse through GH issues FYI).

**Comment by sagikazarmark** [comment]
Yup, I'm aware of the read-only mode. What I'm curious about is whether running an instance in read-only mode and running a separate one in "normal" mode (on a separate host) with the same storage backend and cache could cause any issues.

**Comment by milosgajdos** [comment]
I have done this in past in a few test environments and haven't experienced any problems.

Mind you, my experiments were rather short-lived, so I wouldn't consider them a proper recommendation let alone a benchmark for anything.

Purely speaking from the code perspective, the `read-only` mode leads to `registry` returning `405` for any write operations attempted on registry content, so you should be fine if that's what you are after.

**Comment by sagikazarmark** [comment]
Yeah, when users are not supposed to push to a registry, I don't mind if they receive an error for trying. :)

Is it possible to disable things like garbage collection in the read only instances?

The other thing I'm unsure about is the redis cache: I'm not sure how the registry uses redis. Could there be any problem if it's shared between the write and read instances? Should they have separate redis instances?

@milosgajdos appreciate your help BTW.

**Comment by milosgajdos** [comment]
> Is it possible to disable things like garbage collection in the read-only instances?

garbage collection is a process that is triggered independently of running distribution (https://github.com/distribution/distribution/blob/main/registry/root.go#L43). If you don't want to run GC then don't run it -- it's not a process that runs in a background alongside registry by any means.

> The other thing I'm unsure about is the redis cache: I'm not sure how the registry uses redis.

Regsitry uses Redis to store [blob descriptors](https://github.com/distribution/distribution/blob/main/blobs.go#L61-L86) in order to avoid roundtrips to whatever backing store the raw data is stored in. I can't see any reason why sharing the Redis instance would cause any problems.

**Comment by sagikazarmark** [comment]
Thanks for clarifying. For some reason, I thought GC was running in the background automatically.

**Comment by bigtfromaz** [comment]
@sagikazarmark 

This seems like your idea would be useful in my scenario. We want to PUSH and PULL internally using one container and an internal auth domain. We also want to allow our clients to PULL from the edge using a different auth domain issuing credentials. This will allow us to issue new images to authorized users, while denying them the opportunity to change our repository contents. We'll probably put the edge servers in the cloud and the internal server on-premise.

Were you able to find time to test this out?

**Comment by sagikazarmark** [comment]
@bigtfromaz I haven't, but in theory, it should be possible.

Also, it sounds like you not only want to separate read/write instances, but also want to integrate custom authorization mechanisms.

The registry has a custom auth specification that allows you to do that: https://github.com/distribution/distribution/tree/main/docs/spec/auth

This repo provides a reference implementation: https://github.com/distribution/distribution/tree/main/contrib/token-server

I'm working on a more robust solution though, so if you can wait a few days/weeks, I'll be able to share it with you.

**Comment by bigtfromaz** [comment]
@sagikazarmark I read the above but I'm not sure I would want to implement a token server as opposed to wiring up to a third party implementation, i.e., Azure AD. At this time, we have a well-known population of PULL-only consumers. To keep it simple, I'll simply use two Tomcat basic auth datasets having strong passwords, one used by on-premise container and the other on edge containers. With that said, OpenID Authorization Flow would be desirable down the road. 

I am curious to see what you com up with.

**Comment by sagikazarmark** [comment]
@bigtfromaz it's not completely ready yet, but I decided to publish it: https://github.com/distribution-auth/auth
